router/src/connection_pool.rs

```diff
@@ -0,0 +1,152 @@
+///
+/// This avoids:
+/// 1. That every [`Router`](crate::router::Router) uses their own connections
+/// 2. That we open too many connections in total.
+#[derive(Debug)]
+pub struct ConnectionPool {
+    grpc_clients: LoadingCache<String, Arc<dyn GrpcClient>, ConnectionError>,
+    write_buffer_producers:
+        LoadingCache<KeyWriteBufferProducer, Arc<dyn WriteBufferWriting>, ConnectionError>,
+}
+
+impl ConnectionPool {
+    /// Create new connection pool.
+    ///
+    /// If `use_mock_grpc` is set only mock gRPC clients are created.
+    pub async fn new(use_mock_grpc: bool, wb_factory: WriteBufferConfigFactory) -> Self {
+        // Note: this function is async even though it does not contain any `.await` calls because `LoadingCache::new`
+        // requires tokio to be running and even if documented people will forget about this.
+
+        let grpc_clients = if use_mock_grpc {
+            LoadingCache::new(|_connection_string: String| async move {
+                use crate::grpc_client::MockClient;
+
+                Ok(Arc::new(MockClient::default()) as Arc<dyn GrpcClient>)
+            })
+        } else {
+            LoadingCache::new(|connection_string: String| async move {
+                use crate::grpc_client::RealClient;
+                use influxdb_iox_client::connection::Builder;
+
+                let connection = Builder::default()
+                    .build(&connection_string)
+                    .await
+                    .map_err(|e| Arc::new(e) as ConnectionError)?;
+                Ok(Arc::new(RealClient::new(connection)) as Arc<dyn GrpcClient>)
+            })
+        };
+
+        let wb_factory = Arc::new(wb_factory);
+        let write_buffer_producers = LoadingCache::new(move |key: KeyWriteBufferProducer| {
+            let wb_factory = Arc::clone(&wb_factory);
+            async move {
+                wb_factory
+                    .new_config_write(&key.0, &key.1)
+                    .await
+                    .map_err(|e| Arc::new(EWrapper(e)) as ConnectionError)
+            }
+        });
+
+        Self {
+            grpc_clients,
+            write_buffer_producers,
+        }
+    }
+
+    /// Create new connection factory for testing purposes.
+    #[cfg(test)]
+    pub async fn new_testing() -> Self {
+        use time::SystemProvider;
+
+        let time_provider = Arc::new(SystemProvider::new());
+        Self::new(true, WriteBufferConfigFactory::new(time_provider)).await
+    }
+
+    /// Get gRPC client given a connection string.
+    pub async fn grpc_client(
+        &self,
+        connection_string: &str,
+    ) -> Result<Arc<dyn GrpcClient>, ConnectionError> {
+        let res = self
+            .grpc_clients
+            .get_with_meta(connection_string.to_string())
+            .await
+            .map_err(|e| Arc::new(e) as ConnectionError)?;
+        debug!(was_cached=%res.cached, %connection_string, "getting IOx write client");
```

Comment:
```suggestion
        debug!(was_cached=%res.cached, %connection_string, "getting IOx client");
```

I think the client will be used for more than just `write`, won't it?